### PR TITLE
[feature](mtmv)Support agg state roll up and optimize the roll up code

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes
agg_state is agg  intermediate state, detail see  [state combinator](https://doris.apache.org/zh-CN/docs/dev/sql-manual/sql-functions/combinators/state)

this support agg function roll up as following
 
 <meta charset="utf-8"><div data-page-id="KOvCdvhfko9DqlxchM3cS95xnGh" data-docx-has-block-data="true"><div>
query | materialized view | roll up
-- | -- | --
agg_funtion() | agg_funtion_unoin()  or agg_funtion_state() | agg_funtion_merge()
agg_funtion_unoin() | agg_funtion_unoin() or agg_funtion_state() | agg_funtion_union()
agg_funtion_merge() | agg_funtion_unoin() or agg_funtion_state() | agg_funtion_merge()

for example which can be rewritten by mv sucessfully as following
**mv def is**

```
            select
            o_orderstatus,
            l_partkey,
            l_suppkey,
            sum_union(sum_state(o_shippriority)),
            group_concat_union(group_concat_state(l_shipinstruct)),
            avg_union(avg_state(l_linenumber)),
            max_by_union(max_by_state(l_shipmode, l_suppkey)),
            count_union(count_state(l_orderkey)),
            multi_distinct_count_union(multi_distinct_count_state(l_shipmode))
            from lineitem
            left join orders
            on lineitem.l_orderkey = o_orderkey and l_shipdate = o_orderdate
            group by
            o_orderstatus,
            l_partkey,
            l_suppkey;
```

**query is**

```
            select
            o_orderstatus,
            l_suppkey,
            sum(o_shippriority),
            group_concat(l_shipinstruct),
            avg(l_linenumber),
            max_by(l_shipmode,l_suppkey),
            count(l_orderkey),
            multi_distinct_count(l_shipmode)
            from lineitem
            left join orders 
            on l_orderkey = o_orderkey and l_shipdate = o_orderdate
            group by
            o_orderstatus,
            l_suppkey;
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: regression-test/data/nereids_rules_p0/mv/agg_with_roll_up/aggregate_with_roll_up.out [hunk 1]
```diff
@@ -355,3 +355,51 @@
...
+
+-- !query34_0_after --
+o	3	\	       	
o,o,o,o,o,o,	
              	   mi�      	       	���"K?��D

ZW�_-�A�Vʧ��t�E
+o	4	       	o,o,	              	   yy\r      	       	���"K?��D

ZW
```
Comment (morrySnow) [inline regression-test/data/nereids_rules_p0/mv/agg_with_roll_up/aggregate_with_roll_up.out:380]: agg state result is stable? i don't think so

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java [hunk 2]
```diff
@@ -72,92 +61,15 @@
  */
 public abstract class AbstractMaterializedViewAggregateRule extends AbstractMaterializedViewRule {
 
-    protected static final Multimap<Function, Expression>
-            AGGREGATE_ROLL_UP_EQUIVALENT_FUNCTION_MAP = ArrayListMultimap.create();
+    public static List<AggFunctionRollUpHandler> ROLL_UP_HANDLERS =
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java:64]: final?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java [hunk 3]
```diff
@@ -72,92 +61,15 @@
...
+            Lists.newArrayList(DirectRollupHandler.INSTANCE,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java:65]: use ImmutableList

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java [hunk 4]
```diff
@@ -375,35 +287,22 @@ private boolean isGroupByEquals(Pair<Plan, LogicalAggregate<Plan>> queryTopPlanA
...
-            viewRollupFunction = queryAggregateFunctionShuttled;
-        } else {
-            // handle complex functions roll up
-            // eg: query is count(distinct param), mv sql is bitmap_union(to_bitmap(param))
-            for (Expression mvExprShuttled : mvExprToMvScanExprQueryBased.keySet()) {
-                if (!(mvExprShuttled instanceof Function)) {
+        for (Map.Entry<Expression, Expression> expressionEntry : mvExprToMvScanExprQueryBased.entrySet()) {
+            for (AggFunctionRollUpHandler rollUpHandler : ROLL_UP_HANDLERS) {
+                Pair<Expression, Expression> mvExprToMvScanExprQueryBasedPair = Pair.of(expressionEntry.getKey(),
+                        expressionEntry.getValue());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewAggregateRule.java:293]: init out of inner `for` for better performance

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by starocean999 — APPROVED** [review]

**Pull request merged**